flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -535,41 +545,82 @@ protected void beforeInvoke() throws Exception {
+
+                            IndexedInputGate[] inputGates = getEnvironment().getAllInputGates();
+                            channelIOExecutor.execute(
+                                    () -> {
+                                        try {
+                                            reader.readInputData(inputGates);
+                                        } catch (Exception e) {
+                                            asyncExceptionHandler.handleAsyncException(
+                                                    "Unable to read channel state", e);
+                                        }
+                                    });
+
+                            List<CompletableFuture<?>> recoveredFutures =
+                                    new ArrayList<>(inputGates.length);
+                            for (InputGate inputGate : inputGates) {
+                                recoveredFutures.add(inputGate.getStateConsumedFuture());
+
+                                inputGate
+                                        .getStateConsumedFuture()
+                                        .thenRun(
+                                                () ->
+                                                        mainMailboxExecutor.execute(
+                                                                inputGate::requestPartitions,
+                                                                "Input gate request partitions"));
+                            }
 
-                    for (InputGate inputGate : getEnvironment().getAllInputGates()) {
-                        inputGate
-                                .getStateConsumedFuture()
-                                .thenRun(
-                                        () ->
-                                                mainMailboxExecutor.execute(
-                                                        inputGate::requestPartitions,
-                                                        "Input gate request partitions"));
-                    }
-                });
+                            return gatesRecoveredFuture(suspendableLoop, recoveredFutures);
+                        });
+
+        // Run mailbox until all gates will be recovered.
+        do {
+            suspendableLoop.run();
+        } while (isMailboxLoopRunning() && !allGatesRecoveredFuture.isDone());
+
+        if (canceled) {
+            throw new CancelTaskException();
+        }
+
+        if (!allGatesRecoveredFuture.isDone()) {
+            throw new Exception("Mailbox loop interrupted before recovery was finished.");
+        }
+
+        restored = true;
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> gatesRecoveredFuture(
+            SuspendableMailboxLoop suspendableLoop, List<CompletableFuture<?>> recoveredFutures) {
+        return CompletableFuture.allOf(recoveredFutures.toArray(new CompletableFuture[0]))
+                .thenRun(suspendableLoop::suspend);
+    }
```

Comment:
inline? I don't think it's being used anywhere else?

Reply:
It is using in the test but anyway I am going to rewrite tests in order to avoiding the mockito and then I will be able to inline this method